src/profile/helpers/mutationEditor.ts

```diff
@@ -0,0 +1,520 @@
+    }
+    return list;
+  }
+  if (dataType === 'addresses') {
+    const list: AddressNode[] = getAddressesFromNode(myProfileQuery);
+    const primary = profile.primaryAddress;
+    if (primary) {
+      list.unshift({ ...primary });
+    }
+    return list;
+  }
+  if (dataType === basicDataType) {
+    const { firstName, nickname, lastName, id } = profile;
+    return [
+      {
+        id,
+        firstName,
+        nickname,
+        lastName,
+      },
+    ];
+  }
+  if (dataType === additionalInformationType) {
+    const { language: profileLanguage } = profile;
+    return [
+      {
+        id: '',
+        profileLanguage:
+          profileLanguage || (profileConstants.LANGUAGES[0] as Language),
+      },
+    ];
+  }
+  return [];
+}
+
+export function createEditData(
+  myProfileQuery: ProfileRoot | undefined,
+  dataType: EditData['dataType']
+): EditData[] {
+  const profile = myProfileQuery && myProfileQuery.myProfile;
+  if (!profile) {
+    return [];
+  }
+  const targetData = pickProfileData(myProfileQuery as ProfileRoot, dataType);
+  return targetData.map(targetProfileData =>
+    createEditItem(dataType, targetProfileData)
+  );
+}
+
+export function createEditDataValueFromProfileData(
+  profileDataItem: EditData['profileData'],
+  dataType: EditData['dataType']
+): EditData['value'] {
+  if (dataType === 'phones') {
+    return (profileDataItem as PhoneNode).phone || '';
+  }
+  if (dataType === 'emails') {
+    return (profileDataItem as EmailNode).email || '';
+  }
+  if (dataType === 'addresses') {
+    const {
+      postalCode,
+      address,
+      city,
+      countryCode,
+      primary,
+    } = profileDataItem as AddressNode;
+    return {
+      postalCode,
+      address,
+      city,
+      countryCode,
+      primary,
+    };
+  }
+  if (dataType === basicDataType) {
+    const { firstName, nickname, lastName } = profileDataItem as BasicData;
+    return {
+      firstName,
+      nickname,
+      lastName,
+    };
+  }
+  if (dataType === additionalInformationType) {
+    const { profileLanguage } = profileDataItem as AdditionalInformation;
+    return {
+      id: '',
+      profileLanguage,
+    };
+  }
+  return '';
+}
+
+export function createEditItem(
+  dataType: EditData['dataType'],
+  targetProfileData: EditData['profileData']
+): EditData {
+  return {
+    profileData: targetProfileData,
+    value: createEditDataValueFromProfileData(targetProfileData, dataType),
+    primary: isPrimary(targetProfileData, dataType),
+    dataType,
+  };
+}
+
+function isPrimary(
+  profileDataItem: EditData['profileData'],
+  dataType: EditData['dataType']
+): boolean {
+  if (dataType === 'phones') {
+    return (profileDataItem as PhoneNode).primary === true;
+  }
+  if (dataType === 'emails') {
+    return (profileDataItem as EmailNode).primary === true;
+  }
+  if (dataType === 'addresses') {
+    return (profileDataItem as AddressNode).primary === true;
+  }
+  return false;
+}
```

Comment:
So far I have only been looking at the first commit in this PR, which adds this one file. This comment is about the whole file, although I marked this one function within the file.

I find this whole module doing things in a convoluted way. There are many functions which take in some data and a data type. Then within the function body there's a long `if..else if` structure that checks the data type and does something according to it.

Typescript supports polymorphism and that's exactly a tool to avoid these kinds of long `if..else if`s. As an example, this `isPrimary` function. There could for example exist an interface, let's call it `Primaryable` (please come up with a better name). It has one accessor (field, function, whatever you have available in TS), `primary`. The phone, email and address types implement the `Primaryable` interface. Now you can call `data.primary` for any of these types directly and this `isPrimary` function isn't needed at all anymore.

Repeat the same pattern for other "features" these data types support, and I think you can get rid of most of the functions in this module.

Reply:
You mean:
```
interface Data = {
data:<"some MyProfileNode">
}

interface Primaryable extends Data {
isPrimary():boolean
setPrimary():Primaryable
}

class Address implements Primaryable {
isPrimary(){return this.data.primary}
}
```
"isPrimary" could also be "primary" and a const.

Reply:
For example something like that. Or if Typescript supports implementing multiple interfaces (I don't know if it does), then `Primaryable` shouldn't extend `Data`, but `Address` would implement both `Data` and `Primaryable`.

Reply:
This pattern is for OOP and mainly classes. Current code is functional and there are no classes or objects that have own functions and access to "this". 

In functional programming (FP) just pass the object to a function (usually pure) and it returns what is needed. 

FP: isPrimary(data); // (x)=>return x.primary
OOP: data.isPrimary(); // ()=>return this.primary;

In OOP, instance must also be created to have "isPrimary" and "this". With FP, need for neither exists.

I prefer functional programming over classes. One function handles all scenarios, instead of making multiple functions to each instance. More verbose, but easier to test. Typescript makes it more verbose as it does not allow returning objectX.primary as a boolean if all possible objectXs dont have primary, like basicData.

I can add new interfaces that might save a few lines of code and if-spaghetti.